[JIT] Allow mutated values as functional graph inputs

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35421 [JIT] Incremental updates to Alias Db in Mutation Remover pass
* **#33297 [JIT] Allow mutated values as functional graph inputs**

Allowing mutated values as inputs but not outputs has the effect of buffering up all mutated values as inputs to the graph. Just as we values which escape scope as graph inputs but not graph outputs - we should also allow values that get mutated. In both cases, the contract is that that the functional graph cannot write to graph inputs.

Without this patch, if there is a single write to the Tensor wildcard set it would disable all optimization.

Differential Revision: [D20607175](https://our.internmc.facebook.com/intern/diff/D20607175)

Commit:
Allow mutated values as inputs but not outputs

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow mutated values as functional graph inputs"


Allowing mutated values as inputs but not outputs has the effect of buffering up all mutated values as inputs to the graph. Just as we values which escape scope as graph inputs but not graph outputs - we should also allow values that get mutated. In both cases, the contract is that that the functional graph cannot write to graph inputs.

Without this patch, if there is a single write to the Tensor wildcard set it would disable all optimization.

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow mutated values as functional graph inputs"


Allowing mutated values as inputs but not outputs has the effect of buffering up all mutated values as inputs to the graph. Just as we values which escape scope as graph inputs but not graph outputs - we should also allow values that get mutated. In both cases, the contract is that that the functional graph cannot write to graph inputs.

Without this patch, if there is a single write to the Tensor wildcard set it would disable all optimization.

Commit:
Update on "[JIT] Allow mutated values as functional graph inputs"


Allowing mutated values as inputs but not outputs has the effect of buffering up all mutated values as inputs to the graph. Just as we values which escape scope as graph inputs but not graph outputs - we should also allow values that get mutated. In both cases, the contract is that that the functional graph cannot write to graph inputs.

Without this patch, if there is a single write to the Tensor wildcard set it would disable all optimization.

Commit:
Update on "[JIT] Allow mutated values as functional graph inputs"


Allowing mutated values as inputs but not outputs has the effect of buffering up all mutated values as inputs to the graph. Just as we values which escape scope as graph inputs but not graph outputs - we should also allow values that get mutated. In both cases, the contract is that that the functional graph cannot write to graph inputs.

Without this patch, if there is a single write to the Tensor wildcard set it would disable all optimization.

Commit:
Update on "[JIT] Allow mutated values as functional graph inputs"


Allowing mutated values as inputs but not outputs has the effect of buffering up all mutated values as inputs to the graph. Just as we values which escape scope as graph inputs but not graph outputs - we should also allow values that get mutated. In both cases, the contract is that that the functional graph cannot write to graph inputs.

Without this patch, if there is a single write to the Tensor wildcard set it would disable all optimization.

suo commented on torch/csrc/jit/passes/create_functional_graphs.cpp:
```diff
@@ -159,11 +157,7 @@ struct FunctionalGraphSlicer {
     }
 
     // mutated_values_ already populated with inputs to this node
```
don't need this comment anymore

suo commented on torch/csrc/jit/ir/alias_analysis.h:
```diff
@@ -80,6 +80,9 @@ class AliasDb {
   TORCH_API bool hasWriters(const Node* n) const;
 
+  // Does the node write to any of its inputs
+  TORCH_API bool writeOccursAtNode(Node* n) const;
```
Can we compose this out of existing pieces we have, so we can avoid adding another method? I see a prviate `getWrites`, which is very similar, and from the name `isMutable` does the same job as well.

eellison replied:
yea, i think this was accidentally left in the stack. as you pointed it out it reimplements `writesToAlias`

suo reviewed (approved):
the logic seems reasonable to me, just minor API comments

Commit:
Update on "[JIT] Allow mutated values as functional graph inputs"


Allowing mutated values as inputs but not outputs has the effect of buffering up all mutated values as inputs to the graph. Just as we values which escape scope as graph inputs but not graph outputs - we should also allow values that get mutated. In both cases, the contract is that that the functional graph cannot write to graph inputs.

Without this patch, if there is a single write to the Tensor wildcard set it would disable all optimization.

Differential Revision: [D20607175](https://our.internmc.facebook.com/intern/diff/D20607175)

[ghstack-poisoned]
